### PR TITLE
Ports jni to the proposed jni-sys 0.4 API

### PR DESCRIPTION
**Pull request by rib**
This ports to the `jni-sys` changes I have currently opened as PRs here: https://github.com/jni-rs/jni-sys/pull/28 + https://github.com/jni-rs/jni-sys/pull/23

I'll leave this as a 'draft' until hopefully landing + releasing those `jni-sys` changes.

The two notable changes in jni-sys 0.4 are that JNINativeInterface is now a union that namespaces the functions by the versions they are valid for like:

```rust
struct JNINativeInterface__1_1 {
    GetVersion:
}
struct JNINativeInterface__1_2 {}
struct JNINativeInterface__1_3 {}
struct JNINativeInterface__reserved {}
union JNINativeInterface {
    v1_1: JNINativeInterface__1_1,
    v1_2: JNINativeInterface__1_2,
    v1_3: JNINativeInterface__1_3,
    reserved: JNINativeInterface__reserved,
}
```

and `jboolean` is now an alias for `bool` instead of `u8`

This was used as an opportunity to quite significantly rework the low-level macros used to make jni-sys function calls to make them much simpler and also safer to use.

Key changes:
- The JNI API version needs to be specified along with every function call and will fail to compile if the function doesn't exist for the given version.
- No hidden use of `return` in the macros that can lead to very surprising / non-obvious control flow. Macros evaluate to a single value that may be a Result such that we can use with `?` outside of the macro for early returns.
- The `catch` macro is removed since we don't have to worry about how some macros might use `return` in confusing ways.
- No hidden use of `unsafe` around the most dangerous code! These macros are all very unsafe to use and all callers need to have their own `unsafe` blocks and should ideally document how safety has be considered.
- An infallible JNI function, like GetVersion, can now be called without getting a redundant `Result` to unwrap.
- The macros take `JNIEnv` and `JavaVM` arguments instead of raw pointers which means we don't need `null` pointer checks for each function call.
- Adds more `// Safety:` comments
- Added missing null checks for a number of function calls (e.g. is_assignable_from() and is_instance_of())

The version namespacing has highlighed several places where the jni crate was not checking the version appropriately.

This now makes it so JNI 1.2 is the minimum supported version of JNI since we couldn't even call ExceptionCheck without JNI 1.2

The following functions are now infallible:
  - `JNIEnv::get_version`
  - `JNIEnv::exception_check`
  - `JNIEnv::exception_clear`
  - `JNIEnv::exception_describe`
  - `JNIEnv::is_same_object`
  - `JNIEnv::delete_local_ref`
  - `WeakRef::is_same_object`
  - `WeakRef::is_weak_ref_to_same_object`
  - `WeakRef::is_garbage_collected`

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by [automated tests](https://github.com/jni-rs/jni-rs/blob/master/CONTRIBUTING.md#tests)
- [x] The [coding guidelines](https://github.com/jni-rs/jni-rs/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [x] Public API has documentation
- [x] User-visible changes are mentioned in the Changelog
- [x] The continuous integration build passes

Fixes: #461
Fixes: #463

### CODE REVIEW

**Comment by argv-minus-one** [comment]
I started reviewing this PR, but there's a lot to see here and I've run out of time today. Sorry about that. I'll finish looking this over when I get a chance.

**Comment by rib** [comment]
> I started reviewing this PR, but there's a lot to see here and I've run out of time today. Sorry about that. I'll finish looking this over when I get a chance.

Yep, no worries - thanks for taking a look.

Yeah it's a big change and I'm not sure what a good way of breaking it up would be.

I could have perhaps avoided some of the API changes at the same time, but they are pretty closely related I think.

No big rush, but would be really appreciated to have another pair of eyes look through the changes when you get a chance.

**Review by argv-minus-one — APPROVED** [review]
Good work. This looks like a considerable improvement in both safety and ergonomics. I'm mostly fine with this being merged, although my comment about `ensure_local_capacity` in particular should be addressed before the next release.

File: src/wrapper/java_vm/vm.rs
Comment (rib) [reply src/wrapper/java_vm/vm.rs:138]: I avoided implementing Copy for now just to keep back the possibility of JavaVM becoming a non-transparent wrapper

File: src/wrapper/jnienv.rs
Comment (rib) [reply src/wrapper/jnienv.rs:255]: Yeah, I'd noticed this one too. I think I made an issue to remind us to remove one of them.

ah yep: https://github.com/jni-rs/jni-rs/issues/474

File: src/wrapper/jnienv.rs
Comment (rib) [reply src/wrapper/jnienv.rs:282]: Right, I'd been thinking of just having a wrapper for a `u32` that could have `.major()` and `.minor()` getters that would just pull the upper or lower 16 bits (or conservatively could pull the lower 8 bits and the 3rd byte, while 255 is more than enough head room for now and that avoids assuming nothing weird will be done with the other bits)

File: src/wrapper/errors.rs
Comment (rib) [reply src/wrapper/errors.rs:70]: Yeah, I can tweak this error message to improve it as suggested.

For the bytebuffer APIs they now do runtime version checks with:

`self.ensure_version(JNI_VERSION_1_4)?;`

but it might be better to just set a higher bar - requiring JNI 1.4 should be completely reasonable too.

File: src/wrapper/jnienv.rs
Comment (rib) [reply src/wrapper/jnienv.rs:242]: Maybe - I mainly just wanted it for the internal use case with `MonitorGuard` where we know we've already checked for a null pointer and checked the version - and don't want to unwrap a Result.

I guess similar use cases might potentially arise outside of the crate too though, so maybe it'd be reasonable to make public.

**Comment by argv-minus-one** [comment]
I just realized another problem: the version check in `JNIEnv::from_raw` never happens when the JVM directly calls a Rust function.  `JNIEnv` is `repr(transparent)` and intended to be the first parameter to such functions.

Perhaps we should have a macro attribute for functions intended to be called by the JVM, and `JNIEnv` should *not* be `repr(transparent)`. This would allow us to have whatever initialization and cleanup code we need, including a JNI version check and catching panics. Seems like a lot of work, though, and I don't know how to write macro attributes.

**Comment by argv-minus-one** [comment]
More fun with the `JNIEnv::from_raw` version check. `JavaVM::get_env` and `JavaVM::attach_current_thread*` generate a warning from the JVM if called while an exception is pending:

```
WARNING in native method: JNI call made with exception pending
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x8cff33]  functionEnter(JavaThread*)+0xa3
V  [libjvm.so+0x8d8df4]  checked_jni_GetVersion+0x64
```

The reason is that they call `JNIEnv::from_raw`, which calls C JNI `GetVersion`, which is *not* on [the list of functions](https://docs.oracle.com/en/java/javase/11/docs/specs/jni/design.html#exception-handling) that are safe to call when an exception is pending. (Why it's not on the list, I cannot fathom.)

The `GlobalRef` destructor calls `JavaVM::attach_current_thread`, so this will happen if a `GlobalRef` is dropped while returning early from a Java exception.

Perhaps a sequence of `ExceptionOccurred; ExceptionClear; GetVersion; Throw; DeleteLocalRef` could be used to sidestep this problem—all of those functions are in JNI 1.1—but not without making `JNIEnv::from_raw` surprisingly costly.

**Comment by rib** [comment]
ah, right, fun. Maybe the `GlobalRef` destructor can use a private, exception-safe API for attaching without any version check (via `from_raw_unchecked()`)

It would maybe be neat if there would be a nice type-safe way of us saying we want to attach a thread and only need to access 1.1 APIs and could get back a `JNIEnv_1_1` of some kind that couldn't accidentally be used to access >1.1 APIs.

Or maybe more there could be some way of getting a `JNIEnvSafe` of some kind (either via `get_env()` or fallback to attachment) that would be a limited `JNIEnv` that only exposed exception-safe APIs - and that would never need to do a version check.

File: src/wrapper/java_vm/vm.rs
Comment (rib) [reply src/wrapper/java_vm/vm.rs:355]: I've created a separate issue to remind us to revisit this: https://github.com/jni-rs/jni-rs/issues/500

File: src/wrapper/jnienv.rs
Comment (rib) [reply src/wrapper/jnienv.rs:2964]: Using a `jint` would be a more literal/raw binding of the underlying C JNI API but it implicitly doesn't really make sense to be able to ask for a negative capacity so it seems reasonable to deviate from the C API here for ergonomics.

I don't think it would generally be realistic to even expect that values near to `i32::MAX` would be likely to succeed, the choice to use `usize` here is really just because it seems like the appropriate the Rusty type for specifying this kind of size.

Given the choice between `u32` or `usize` then we'd still have to account for a value potentially > `i32::MAX` either way but `usize` feels more appropriate here.

If we have cases like negative values that could return an error without necessarily throwing an OutOfMemory exception then it seems like anyway the caller can't really just assume that an OutOfMemory exception is the only failure mode here.

It probably would have been better to change separately though.

Sort of inclined to leave this atm, unless you feel strongly about this?

Comment (argv-minus-one) [reply src/wrapper/jnienv.rs:2964]: I took another look at this, and have changed my mind.

[HotSpot also returns an error if a bogus maximum capacity is given](https://github.com/openjdk/jdk/blob/09dad0e96b37e3fcd1a13040e0de85ebc04b07c2/src/hotspot/share/prims/jni.cpp#L725), and [the default upper limit is well below `i32::MAX`](https://github.com/openjdk/jdk/blob/09dad0e96b37e3fcd1a13040e0de85ebc04b07c2/src/hotspot/share/runtime/globals.hpp#L781), so this proposed behavior is consistent with HotSpot.

It probably should be noted in this method's documentation that an excessively large capacity will result in an error, and that HotSpot's default maximum capacity is 65536.

File: src/wrapper/macros.rs
Comment (rib) [reply src/wrapper/macros.rs:73]: Currently it can't be a function because there's no trait that corresponds to the `is_null()` check for all the types that may be passed to `null_check`. We could potentially create one function that accepts an `AsRef<JObject<'a>>` but would then need separate APIs for non-object / raw pointer types. E.g it's also used like this currently:

```rust
pub unsafe fn from_raw(ptr: *mut sys::JavaVM) -> Result<Self> {
        let ptr = null_check(ptr, "from_raw ptr argument")?;
```

File: src/wrapper/errors.rs
Comment (rib) [reply src/wrapper/errors.rs]: For reference here; I bumped the version checked in `JNIEnv::from_raw()` to 1.4 so the runtime checks for 1.4 could be removed for the direct byte buffer APIs.

File: src/wrapper/jnienv.rs
Comment (rib) [reply src/wrapper/jnienv.rs]: For reference here; I ended up making this `pub`

File: src/wrapper/jnienv.rs
Comment (rib) [reply src/wrapper/jnienv.rs]: For reference here:

I updated the `JNIVersion` type to use more conventional JNI version designations (so `V1`, `V2`, `V4`, `V6` and `V8` are now `V1_1`, `V1_2`, `V1_4`, `V1_6` and `V1_8`) and the type is now a transparent `i32` wrapper that just provides a more ergonomic way to check a raw JNI version. There is no 'invalid' version any more so even if `JNIVersion` has not been updated to know about some future `V30` it is still possible to check the `.major()` and `.minor()` component and it is valid to do `Ord`ered comparisons between versions.

The public `JNIEnv::get_version()` API was renamed to `JNIEnv::version()` since that's more idiomatic for a Rust API.

File: src/wrapper/jnienv.rs
Comment (rib) [reply src/wrapper/jnienv.rs:3218]: For reference here; I added a dep on this crate and assert that `MonitorGuard` doesn't implement `Send`

File: src/wrapper/java_vm/vm.rs
Comment (rib) [reply src/wrapper/java_vm/vm.rs]: Since I ended up making `get_env` changes to mark it unsafe and use `JNIEnv::from_raw_unchecked()` I took the opportunity to also have it accept a `version` argument.

The safety docs for `get_env` now require the caller to know that the `JavaVM` supports at least JNI 1.4.

**Comment by rib** [comment]
> More fun with the `JNIEnv::from_raw` version check. `JavaVM::get_env` and `JavaVM::attach_current_thread*` generate a warning from the JVM if called while an exception is pending

For reference; I've ended up making a few changes here:
1. `get_env` is now unsafe, and accepts a `version` argument that must be >= `V1_4`. The safety docs make it the callers responsibility to know that the `JavaVM` supports >= JNI 1.4.
2. `get_env` uses `JNIEnv::from_raw_unchecked` now.

Marking `get_env` unsafe also highlights the lack of safety from being able to potentially materialize a mutable `JNIEnv` with a lifetime that might not correspond to the top stack frame for local object references. (re: https://github.com/jni-rs/jni-rs/discussions/436#discussioncomment-5421738)

**Comment by rib** [comment]
Okey, I think I've addressed all of your feedback @argv-minus-one (thanks). Considering this is a pretty wide-spread change that is holding up addressing various smaller issues I'm going to merge this now and we can hopefully deal with any further follow ups via separate issues / PRs.

**Pull request merged**
